### PR TITLE
Add final project implementation from Assignment 2

### PR DESCRIPTION
**Pull request by FLSGawsl**

### CODE REVIEW

File: src/main/java/StringManipulation.java [hunk 1]
```diff
@@ -1,33 +1,82 @@
...
-    public String[] getSubStrings(int startWord, int endWord) {
-        return null;
+    public String[] getSubStrings(int startWord, int endWord){
+        if (string == null) {
+            return null;
+        }
+        String[] words = string.trim().split("\\s+");
+        int numWords = words.length;
+
+        if (startWord <= 0 || endWord <= 0 || startWord > endWord) {
```
Comment (brunochristensen) [inline src/main/java/StringManipulation.java:56]: This might be more of a style choice, but wouldn't it make more sense to throw exceptions before executing business logic?

File: src/main/java/StringManipulation.java [hunk 2]
```diff
@@ -1,33 +1,82 @@
...
     public String restoreString(int[] indices) {
-        return null;
+        if (string == null || indices == null || string.length() != indices.length) {
+            throw new IllegalArgumentException("Invalid arguments: string length must be equal to indices length");
+        }
+        char[] chars = string.toCharArray();
+        char[] restoredChars = new char[chars.length];
+        for (int i = 0; i < indices.length; i++) {
+            int index = indices[i];
+            if (index < 0 || index >= chars.length) {
```
Comment (brunochristensen) [inline src/main/java/StringManipulation.java:74]: Would it make more sense to verify the integrity of your data prior to when you start executing the logic of the method?

File: src/test/java/StringManipulationTest.java [hunk 3]
```diff
@@ -7,142 +7,192 @@
...
-        fail("Not yet implemented");
+        manipulatedString.setString("Holy mama");
+        assertEquals("Hol  ma a", manipulatedString.removeNthCharacter(4, true));
     }
 
     @Test
     public void testRemoveNthCharacter6() {
-        fail("Not yet implemented");
+        manipulatedString.setString("Exception testing1");
+        assertEquals("Don't mind this", manipulatedString.removeNthCharacter(0, true));
```
Comment (brunochristensen) [inline src/test/java/StringManipulationTest.java:83]: JUnit does provide a assertThrows() method to test exception handling. It would probably be better than relying on the exception methods returning string.

File: src/main/java/StringManipulation.java [hunk 4]
```diff
@@ -1,33 +1,82 @@
...
+            return null;
+        }
+        if (n <= 0) {
+            throw new IllegalArgumentException("n must be greater than zero");
+        }
+        if (n > string.length()) {
+            throw new IndexOutOfBoundsException("n exceeds string length");
+        }
+        StringBuilder result = new StringBuilder();
+        char[] chars = string.toCharArray();
```
Comment (brunochristensen) [inline src/main/java/StringManipulation.java:35]: It doesn't look like you ever do anything with the chars variable.

File: src/test/java/StringManipulationTest.java [hunk 5]
```diff
@@ -7,142 +7,192 @@
...
-        manipulatedstring.setString("This is my string");
-        int length = manipulatedstring.count();
+        manipulatedString.setString("This is my string");
+        int length = manipulatedString.count();
         assertEquals(4, length);
     }
 
     @Test
     public void testCount2() {
-        fail("Not yet implemented");
```
Comment (brunochristensen) [inline src/test/java/StringManipulationTest.java:31]: I would consider adding more variety or some edge cases to your testing. Maybe add some sentences with unconventional spacing for characters.

File: src/main/java/StringManipulation.java [hunk 6]
```diff
@@ -1,33 +1,82 @@
...
+        this.string = string;
     }
 
     @Override
     public int count() {
-        return 0;
+        if (string == null) {
+            return 0;
+        }
+        String[] words = string.trim().split("\\s+");
```
Comment (brunochristensen) [inline src/main/java/StringManipulation.java:19]: trim() is probably an unnecessary step to separate word by whitespace.

File: src/test/java/StringManipulationTest.java [hunk 7]
```diff
@@ -7,142 +7,192 @@
...
+        assertEquals(sStings[0], "Why");
+        assertEquals(sStings[1], "there");
+        assertEquals(sStings[2], "are");
     }
     @Test
     public void testGeSubStrings5() {
-        fail("Not yet implemented");
+        manipulatedString.setString("Why there are so many test cases to create");
+        String [] sStings = manipulatedString.getSubStrings(0, 6);
+
```
Comment (brunochristensen) [inline src/test/java/StringManipulationTest.java:130]: How is this testing if an exception is thrown? What are the mechanics behind this?

File: src/test/java/StringManipulationTest.java [hunk 8]
```diff
@@ -7,142 +7,192 @@
...
 
     @Test
     public void testRestoreString2()
     {
-        fail("Not yet implemented");
-
+        manipulatedString.setString("art");
+        int [] array;
+        array=new int[]{2,0,1};
+        String restoreString = manipulatedString.restoreString(array);
```
Comment (brunochristensen) [inline src/test/java/StringManipulationTest.java:162]: What would happen if the array argument uses duplicate values?

**Review by brunochristensen — COMMENTED** [review]
The code looks good and functional, I would just think about expanding the scope of your test cases.
